File: 1-50/problem21/problem21.rb
def divisors_sum(n)
  (2..n**0.5).select { |x| n % x == 0 }.map { |x| x + n / x }.inject(0, &:+) + 1
end

sums = []
1.upto(9999) { |x| sums[x] = divisors_sum(x) }
puts 1.upto(9999).select { |x| sums[sums[x]] == x && sums[x] != x }
      .inject(0) { |sum, x| sum + sums[x] }

File: 1-50/problem21/problem21.scala
def divisors_sum(n: Int) = {
  val sqrt = math.sqrt(n).toInt
  1 + (2 to sqrt).filter(n % _ == 0).map(x => x + n / x).sum
}

val sums = (1 to 9999).map(x => (x, divisors_sum(x))).toMap
println((1 to 9999).filter(x => sums(x) < 10000 && sums(sums(x)) == x && sums(x) != x).map(sums(_)).sum)